hw/drivers/sensors/bma2xx/src/bma2xx.c

```diff
@@ -0,0 +1,4940 @@
+    if (rc != 0) {
+        BMA2XX_ERROR("I2C write failed at address 0x%02X single byte\n",
+                     addr);
+        return rc;
+    }
+
+    return 0;
+}
+
+static int
+get_register(const struct bma2xx * bma2xx,
+             uint8_t addr,
+             uint8_t * data)
+{
+    int rc;
+    const struct sensor_itf * itf;
+    itf = SENSOR_GET_ITF(&bma2xx->sensor);
+
+    if (itf->si_type == SENSOR_ITF_SPI) {
+        rc = spi_readlen(itf, addr, data, 1);
+    } else if (itf->si_type == SENSOR_ITF_I2C){
+        rc = i2c_readlen(itf, addr, data, 1);
+    } else {
+        rc = SYS_EINVAL;
+    }
+
+    return rc;
+}
+
+static int
+get_registers(const struct bma2xx * bma2xx,
+              uint8_t addr,
+              uint8_t * data,
+              uint8_t size)
+{
+    int rc;
+    const struct sensor_itf * itf;
+    itf = SENSOR_GET_ITF(&bma2xx->sensor);
+
+    if (itf->si_type == SENSOR_ITF_SPI) {
+        rc = spi_readlen(itf, addr, data, size);
+    } else if (itf->si_type == SENSOR_ITF_I2C){
+        rc = i2c_readlen(itf, addr, data, size);
+    } else {
+        rc = SYS_EINVAL;
+    }
+
+    return rc;
+}
+
+static int
+set_register(const struct bma2xx * bma2xx,
+             uint8_t addr,
+             uint8_t data)
+{
+    int rc;
+    const struct sensor_itf * itf;
+
+    itf = SENSOR_GET_ITF(&bma2xx->sensor);
+
+    if (itf->si_type == SENSOR_ITF_SPI) {
+        rc = spi_writereg(itf, addr, data, 1);
+    } else if (itf->si_type == SENSOR_ITF_I2C){
+        rc = i2c_writereg(itf, addr, data);
+    } else {
+        rc = SYS_EINVAL;
+    }
+
+    switch (bma2xx->power) {
+    case BMA2XX_POWER_MODE_SUSPEND:
+    case BMA2XX_POWER_MODE_LPM_1:
+        delay_msec(1);
+        break;
+    default:
+        break;
+    }
+
+    return rc;
+}
+
+int
+bma2xx_get_chip_id(const struct bma2xx * bma2xx,
+                   uint8_t * chip_id)
+{
+    return get_register(bma2xx, REG_ADDR_BGW_CHIPID, chip_id);
+}
+
+static void
+compute_accel_data(struct accel_data * accel_data,
+                   enum bma2xx_model model,
+                   const uint8_t * raw_data,
+                   float accel_scale)
+{
+    int16_t raw_accel;
+    uint8_t model_shift = 0;
+
+    switch (model) {
+        case BMA2XX_BMA280:
+            model_shift = BMA280_ACCEL_BIT_SHIFT;
+            break;
+        case BMA2XX_BMA253:
+            model_shift = BMA253_ACCEL_BIT_SHIFT;
+            break;
+        default:
+            break;
+    }
+
+    raw_accel = (int16_t)(raw_data[0] & 0xFC) | (int16_t)(raw_data[1] << 8);
+    raw_accel >>= model_shift;
+
+    accel_data->accel_g = (float)raw_accel * accel_scale;
+    accel_data->new_data = raw_data[0] & 0x01;
+}
+
+static int get_accel_scale(enum bma2xx_model model,
+                            enum bma2xx_g_range g_range,
+                            float *accel_scale)
+{
+    switch (g_range) {
+        case BMA2XX_G_RANGE_2:
+            switch(model){
```

Comment:
maybe instead of switch(model) we could some const table of structs with custom parameters?

Reply:
Yeah, I was having second thoughts about this implementation, and reviewed with @vrahane and @wes3.  Given the limited number of models (7) we agreed it was okay to use this switch.  Is it okay if we keep this as is for now?